From c3a4fefa9194b66753b27d8a8961a5cb5f2360f8 Mon Sep 17 00:00:00 2001
From: John Galt <johngaltfirstrun@gmail.com>
Date: Tue, 28 Jan 2025 13:01:49 -0500
Subject: [PATCH] CameraProviderExtension: add enabled bool and always set
 torch

This has two changes for two reasons:
 - Enabled bool is for device side workarounds (such as switches which must be reset, etc).

 - Always setting torch strength in this path fixes normal flash usage
   (camera flash, screen off torch gestures, etc).

Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 .../libcameraservice/common/CameraProviderExtension.cpp    | 2 +-
 .../libcameraservice/common/CameraProviderExtension.h      | 2 +-
 .../libcameraservice/common/CameraProviderManager.cpp      | 7 +++----
 3 files changed, 5 insertions(+), 6 deletions(-)

diff --git a/services/camera/libcameraservice/common/CameraProviderExtension.cpp b/services/camera/libcameraservice/common/CameraProviderExtension.cpp
index 6dbabbad8d..65435a4d23 100644
--- a/services/camera/libcameraservice/common/CameraProviderExtension.cpp
+++ b/services/camera/libcameraservice/common/CameraProviderExtension.cpp
@@ -34,6 +34,6 @@ __attribute__((weak)) int32_t getTorchStrengthLevelExt() {
     return 0;
 }
 
-__attribute__((weak)) void setTorchStrengthLevelExt(__unused int32_t torchStrength) {
+__attribute__((weak)) void setTorchStrengthLevelExt(__unused int32_t torchStrength, __unused bool enabled) {
     // Nothing
 }
diff --git a/services/camera/libcameraservice/common/CameraProviderExtension.h b/services/camera/libcameraservice/common/CameraProviderExtension.h
index ae17a2fc00..693400b759 100644
--- a/services/camera/libcameraservice/common/CameraProviderExtension.h
+++ b/services/camera/libcameraservice/common/CameraProviderExtension.h
@@ -20,4 +20,4 @@ extern bool supportsTorchStrengthControlExt();
 extern int32_t getTorchDefaultStrengthLevelExt();
 extern int32_t getTorchMaxStrengthLevelExt();
 extern int32_t getTorchStrengthLevelExt();
-extern void setTorchStrengthLevelExt(int32_t torchStrength);
+extern void setTorchStrengthLevelExt(int32_t torchStrength, bool enabled);
diff --git a/services/camera/libcameraservice/common/CameraProviderManager.cpp b/services/camera/libcameraservice/common/CameraProviderManager.cpp
index 643de132ae..b24c3c8d77 100644
--- a/services/camera/libcameraservice/common/CameraProviderManager.cpp
+++ b/services/camera/libcameraservice/common/CameraProviderManager.cpp
@@ -612,7 +612,7 @@ status_t CameraProviderManager::turnOnTorchWithStrengthLevel(const std::string &
         }
 
         deviceInfo->setTorchMode(torchStrength > 0);
-        setTorchStrengthLevelExt(torchStrength);
+        setTorchStrengthLevelExt(torchStrength, true);
         deviceInfo->mTorchStrengthLevel = torchStrength;
         return OK;
     } else {
@@ -712,11 +712,10 @@ status_t CameraProviderManager::setTorchMode(const std::string &id, bool enabled
     saveRef(DeviceMode::TORCH, deviceInfo->mId, halCameraProvider);
 
     res = deviceInfo->setTorchMode(enabled);
-    if (!enabled &&
-        deviceInfo->hasFlashUnit() && supportsTorchStrengthControlExt()) {
+    if (deviceInfo->hasFlashUnit() && supportsTorchStrengthControlExt()) {
         // Need to reset torch strength back to default when torch is turned off
         int32_t defaultLevel = getTorchDefaultStrengthLevelExt();
-        setTorchStrengthLevelExt(defaultLevel);
+        setTorchStrengthLevelExt(defaultLevel, enabled);
         deviceInfo->mTorchStrengthLevel = defaultLevel;
     }
     return res;
-- 
2.53.0

